### ISSUE TITLE
Add a `CollisionShape32D` node

### ISSUE THREAD

**Issue by ghsoares** [godotengine/godot-proposals#2568]
### Describe the project you are working on
A 2D game with some 3D rendering

### Describe the problem or limitation you are having in your project
Creating a project in Godot with 3D renderer but with 2D physics interactions requires some hacks, for example setting manually a RigidBody transform to match a RigidBody2D transform (while converting the 2D pixels unit to 3D unit), using constraints to the rigidbody so that the z position and xy rotation don't change (creates inconsistencies when compared to a fully 2D physics simulation), etc. It would be better (in my opinion) if there was some node to resolve 2D collisions in 3D.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation
The feature that I propose is a `CollisionShape32D` node (or other name that describe a collision shape that resolves 2D collision in 3D). This would be a `Spatial` node with the properties of a `CollisionShape2D` node (shape, disabled, one way collision and one way collision margin), the `shape` property accepts any `Shape2D` type resource just like the `CollisionShape2D` node. 

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams
I'm not sure how the `PhysicsServer` server is implemented, but maybe if the server can handle a `CollisionShape32D` and pass it to the `PhysicsServer2D` server, compute the physics simulation frame, then converting the results from 2D to 3D (while converting the 2D pixels unit to 3D unit, maybe there could be a option in the node to change the pixel size for convertion) to update the parent body's transform, this depends on how the physics back-end is implemented. The conversion between 3D and 2D would require some steps:
- Multiply the spatial coordinates by a factor (pixel size in this case, defaulting to 0.01), divide by this factor when converting to 3D;
- Ignore the Z position/index (when converting 2D to 3D, the Z position of the transform cannot be changed);
- Flip the Y axis (Y up direction in 2D is negative, while being positive in 3D).

### If this enhancement will not be used often, can it be worked around with a few lines of script?
That would require a manual transformation synchronization between a RigidBody and a RigidBody2D, for example, as well could easilly mess with the scene organization, with a bunch of 2D nodes mixed with Spatial nodes.

### Is there a reason why this should be core and not an add-on in the asset library?
Because when on core, it could be optimized to better performance than creating a script to do it. As well it would help to beginners to create a 3D game with 2D physics without needing extra scripts just for this 2D/3D physics interaction.

**Comment by SlugFiller** [comment]
Your approach is wrong. If the physics is truly purely 2d, then your view node shouldn't be a RigidBody, it should be a Spatial. What you really want is model/view separation, allowing you to have 2D physics which translate into a 3D view. Incidentally, this is also important for client/server net games, where you want the physics to run on the server, but the view to occur on the client.

At any rate, there's no straightforward way for the core to support view/model separation. It would basically require a way to listen to all changes in a scene tree, and create a shadow tree. Instead of having such a global listener as a core object, it makes more sense to simply have the individual model objects make calls on "process", "_enter_tree" and "_exit_tree" to update a custom scene-state listener.

**Comment by LightningAA** [comment]
32 dimensions? That might get a bit performance intensive... ;)

**Comment by ghsoares** [comment]
@SlugFiller View/Model separations seens to be the way to go in this feature, this seens to be similar to the first hacky work around to update via script the spatial transforms based on the 2D transforms. If there could be way to implement this on core, so some kind of `RemoteTransform32D` would be easier of implement?

**Comment by ghsoares** [comment]
@LightningAA hahaha 32D is the sugestion that I made to the name of the node, I didn't think of other name that could easilly describe a 2D with 3D interaction or vice-versa.

**Comment by SlugFiller** [comment]
Copying a transform on _process would do the trick. That's how Smoothing Node from the asset library works. I'm not sure how well _process syncs with the VisualServer's refresh rate, but then, the physics isn't perfectly synced to the framerate, either.

The part that's "hard" is spawning a view node every time you spawn a model node. For instance, if your character shoots, you need to simultaneously spawn the 2D bullet collision object, and then the 3D bullet model object, and have the latter reference the prior for its position.

Incidentally, you could completely resolve even that little bit of extra work by using ECS. You simply add a 2d physics component, and a 3d view component, on the same entity. You add a custom component which converts the 2d transform to a 3d one, and done. I haven't personally tested the Godex library, however, so I can't give any direct tips on how to use it.

**Comment by Xrayez** [comment]
By the way, there's https://github.com/98teg/SpriteMesh which allows to create 3D meshes from 2D sprites. I realize that it's marginally related to this proposal, but conceptually this could be extended to provide collisions from those meshes.

**Comment by ghsoares** [comment]
@Xrayez Not exactly, this plugin seems to extrude a sprite to have depth in 3D, but my proposal is quite different, it proposes the ability to use 2D collision shapes in a 3D scene, so anyone can have 2D physics working perfectly fine in a 3D scene without needing to lock a axis from a physics body.
